src/script/lua_api/l_areastore.cpp

```diff
@@ -0,0 +1,420 @@
+{
+	if (!(borders || data)) {
+		lua_pushboolean(L, true);
+	}
+	lua_newtable(L);
+	if (borders) {
+		lua_pushstring(L, "min");
+		push_v3s16(L, a->minedge);
+		lua_settable(L,-3);
+		lua_pushstring(L, "max");
+		push_v3s16(L, a->maxedge);
+		lua_settable(L,-3);
+	}
+	if (data) {
+		lua_pushstring(L, "data");
+		lua_pushlstring(L, a->data, a->datalen);
+		lua_settable(L,-3);
+	}
+}
+
+static inline void push_areas(lua_State *L, const std::vector<Area *> &areas,
+		bool borders, bool data)
+{
+	lua_newtable(L);
+	size_t cnt = areas.size();
+	for (size_t i = 0; i < cnt; i++) {
+		lua_pushnumber(L, areas[i]->id);
+		push_area(L, areas[i], borders, data);
+		lua_settable(L,-3);
+	}
+}
+
+// garbage collector
+int LuaAreaStore::gc_object(lua_State *L)
+{
+	LuaAreaStore *o = *(LuaAreaStore **)(lua_touserdata(L, 1));
+	delete o;
+	return 0;
+}
+
+// get_area(id, borders, data)
+int LuaAreaStore::l_get_area(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	LuaAreaStore *o = checkobject(L, 1);
+	AreaStore *ast = o->as;
+
+	u32 id = lua_tonumber(L, 2);
+
+	bool data = false;
+	if (lua_isboolean(L, 3)) {
+		data = lua_toboolean(L, 3);
+	}
+
+	const Area *res;
+	
+	res = ast->getArea(id);
+	push_area(L, res, true, data);
+
+	return 1;
+}
+
+// get_areas_for_pos(pos, borders, data)
+int LuaAreaStore::l_get_areas_for_pos(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	LuaAreaStore *o = checkobject(L, 1);
+	AreaStore *ast = o->as;
+
+	v3s16 pos = check_v3s16(L, 2);
+
+	bool borders = true;
+	bool data = false;
+	get_data_and_border_flags(L, 3, &borders, &data);
+
+	std::vector<Area *> res;
+	
+	ast->getAreasForPos(&res, pos);
+	push_areas(L, res, borders, data);
+
+	return 1;
+}
+
+// get_areas_in_area(area, accept_overlap, borders, data)
+int LuaAreaStore::l_get_areas_in_area(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	LuaAreaStore *o = checkobject(L, 1);
+	AreaStore *ast = o->as;
+
+	luaL_checktype(L, 2, LUA_TTABLE);
+	lua_getfield(L, 2, "min");
+	v3s16 minedge = check_v3s16(L, -1);
+	lua_pop(L, 1);
+
+	lua_getfield(L, 2, "max");
+	v3s16 maxedge = check_v3s16(L, -1);
+	lua_pop(L, 1);
+
+	bool borders = true;
+	bool data = false;
+	bool accept_overlap = false;
+	if (lua_isboolean(L, 3)) {
+		accept_overlap = lua_toboolean(L, 3);
+		get_data_and_border_flags(L, 4, &borders, &data);
+	}
+	std::vector<Area *> res;
+	
+	ast->getAreasInArea(&res, minedge, maxedge, accept_overlap);
+	push_areas(L, res, borders, data);
+
+	return 1;
+}
+
+// insert_area(area)
+int LuaAreaStore::l_insert_area(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+
+	LuaAreaStore *o = checkobject(L, 1);
+	AreaStore *ast = o->as;
+
+	Area a;
+
+	luaL_checktype(L, 2, LUA_TTABLE);
+	lua_getfield(L, 2, "min");
+	a.minedge = check_v3s16(L, -1);
+	lua_pop(L, 1);
+
+	lua_getfield(L, 2, "max");
+	a.maxedge = check_v3s16(L, -1);
+	lua_pop(L, 1);
+
+	a.extremifyEdges();
+	a.id = ast->getFreeId(a.minedge, a.maxedge);
+
+	if (a.id != 0) {
```

Comment:
If 0 means "no free IDs", that value has a special meaning and should have its own #define.
Also, for this if...else there's barely any work being done in the else (aside from handling an error case).  It might be cleaner to "bail out" on failure.